source/extensions/filters/http/cache/cache_insert_queue.cc

```diff
@@ -0,0 +1,183 @@
+#include "source/extensions/filters/http/cache/cache_insert_queue.h"
+
+#include "source/common/buffer/buffer_impl.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace Cache {
+
+class CacheInsertChunk {
+public:
+  // Sends a chunk to the cache.
+  // on_complete is called when the cache completes the operation.
+  virtual void send(InsertContext& context,
+                    std::function<void(bool ready, bool end_stream, size_t sz)> on_complete) PURE;
+
+  virtual ~CacheInsertChunk() = default;
+};
+
+class CacheInsertChunkBody : public CacheInsertChunk {
+public:
+  CacheInsertChunkBody(const Buffer::Instance& buffer, bool end_stream)
+      : buffer_(buffer), end_stream_(end_stream) {}
+
+  void send(InsertContext& context,
+            std::function<void(bool ready, bool end_stream, size_t sz)> on_complete) override {
+    size_t sz = buffer_.length();
+    context.insertBody(
+        std::move(buffer_),
+        [on_complete, end_stream = end_stream_, sz](bool ready) {
+          on_complete(ready, end_stream, sz);
+        },
+        end_stream_);
+  }
+
+private:
+  Buffer::OwnedImpl buffer_;
+  const bool end_stream_;
+};
+
+class CacheInsertChunkTrailers : public CacheInsertChunk {
+public:
+  explicit CacheInsertChunkTrailers(const Http::ResponseTrailerMap& trailers)
+      : trailers_(Http::ResponseTrailerMapImpl::create()) {
+    Http::ResponseTrailerMapImpl::copyFrom(*trailers_, trailers);
+  }
+
+  void send(InsertContext& context,
+            std::function<void(bool ready, bool end_stream, size_t sz)> on_complete) override {
+    // While zero isn't technically true for the size of trailers, it doesn't
+    // matter at this point because watermarks after the stream is complete
+    // aren't useful.
+    context.insertTrailers(*trailers_, [on_complete](bool ready) { on_complete(ready, true, 0); });
+  }
+
+private:
+  std::unique_ptr<Http::ResponseTrailerMap> trailers_;
+};
+
+CacheInsertQueue::CacheInsertQueue(Http::StreamEncoderFilterCallbacks& encoder_callbacks,
+                                   InsertContextPtr insert_context, AbortInsertCallback abort)
+    : dispatcher_(encoder_callbacks.dispatcher()), insert_context_(std::move(insert_context)),
+      low_watermark_bytes_(encoder_callbacks.encoderBufferLimit() / 2),
+      high_watermark_bytes_(encoder_callbacks.encoderBufferLimit()),
+      encoder_callbacks_(encoder_callbacks), abort_callback_(abort) {}
+
+void CacheInsertQueue::insertHeaders(const Http::ResponseHeaderMap& response_headers,
+                                     const ResponseMetadata& metadata, bool end_stream) {
+  if (end_stream) {
+    end_in_sight_ = true;
+  }
+  // While zero isn't technically true for the size of headers, headers are
+  // typically excluded from the stream buffer limit.
+  insert_context_->insertHeaders(
```

Comment:
This may just be a caching thing so ignore this if so but I associate chunks with HTTP 1.1 chunking and having the header be considered a chunk is really weird.
I'd be inclined to have these all be bytes or data based (data in flight) or something to avoid that, but if considering headers/trailers to be chunks is normal caching nomenclature feel free to leave as is.  cc @jmarantz @RyanTheOptimist for thoughts

Reply:
I don't think it's particularly a caching thing, and it's reasonable to flinch from the name conflict, but e.g. "data in flight" is misleading because "in flight" here (and in async file system stuff) is used to mean stuff that's not waiting in a buffer or queue, whereas this is stuff that *is* waiting in a queue, and saying "data" is almost the same as saying nothing at all. "chunk" indicates that it's a *piece* of what's in the queue, where "data" to me would imply a structure containing all the data. "Slice" would be an option but is used by Envoy::Buffer, so would be a potentially confusing naming conflict there.

A thesaurus doesn't suggest any other great alternatives. Wodge. Lump.